W Staking (#529)
wormhole-foundation/wormhole-docs: protocol/staking/w-staking.md

---
title: W Staking
description: Integrate W staking into your app on EVM and Solana. Learn flows, contract calls, and delegate discovery using the Tally API.
---

# W Staking Integration

> W staking allows users to stake their W tokens to participate in governance and earn staking rewards in return, while retaining control over their tokens. This guide walks you through integrating native W staking for both EVM chains (Ethereum, Optimism, Arbitrum, Base) and Solana, including high-level differences, contract calls, and recommended delegate discovery via the Tally API.

Reviewer: Maybe link to Tally API

### W Token Contract Call

```js 
const tx = await writeContractAsync({
  address: W_TOKEN_ADDRESS,
  abi: ERC1967,
  functionName: 'delegate',
  args: [isStaking ? delegateAddress : zeroAddress],
  chainId: chainId,
})
```

- To stake, pass the `delegateAddress`.
- To unstake, pass the `zeroAddress` (`0x000...000`).

Use the `delegates(address)` view function to check the current delegate for a given address.

!!!note "Important Notes"
    - There can only be one delegate per wallet. 
    - You can delegate to yourself or someone else.
    - Wormhole maintains a list of delegates using the Tally API (covers EVM and Solana).

## Solana Integration

On Solana, staking means moving W into a stake custody account that’s tied to the staker and assigning a delegate for the voting power of the amount you staked. You can stake any amount and you can delegate to yourself or to a third party.

### Prerequisites

- `@solana/web3.js`, `@solana/spl-token`, and `@coral-xyz/anchor` installed.
- A Connection to the target cluster (e.g., mainnet).
- The W token mint on Solana.

> - The Staking program ID (Anchor program) and its IDL.

Reviewer: Link to the anchor program

| `currentDelegatePublicKey` | `PublicKey` | Your current delegate’s address when switching delegates (or `null` if staking for the first time). | `null` |
| `amount` | `BN` or `bigint` | The number of W tokens to stake in base units (respecting the token’s decimals). | `new BN(10 * 10 ** 9)` if W has 9 decimals |

### Staking Flow (Solana)

Initialize the staking program and config. See [Program IDs and ABIs section](/docs/protocol/staking/w-staking/#program-ids-and-abis) for more details.

Staking steps:

1. Ensure the user has a W token ATA (Associated Token Account).
2. Create a stake account if one doesn't already exist.
3. Transfer tokens to the stake custody account.
4. Create a delegate stake account if needed.
5. Send the `delegate()` instruction.

```js   
--8<-- 'code/protocol/staking/staking-flow.js'
```

### Unstaking Flow (Solana)

Use `withdrawTokens(amount, currentDelegate, recipient)` to pull staked W back to your ATA. If you self-delegated, `currentDelegate` is your own key.

```js 
--8<-- 'code/protocol/staking/unstaking-flow.js'
```

> 1. Confirm the user’s stake metadata and custody accounts.

Reviewer: In the section before (Staking Flow), you first show the steps, followed by the code. Here, you first have the code, followed by the steps. Move these steps above the code snippet.

---
title: W Staking
description: Integrate W staking into your app on EVM and Solana. Learn flows, contract calls, and delegate discovery using the Tally API.
---

# W Staking Integration

W staking allows users to stake their W tokens to participate in governance and earn staking rewards in return, while retaining control over their tokens. This guide walks you through integrating native W staking for both EVM chains (Ethereum, Optimism, Arbitrum, Base) and Solana, including high-level differences, contract calls, and recommended delegate discovery via the [Tally API](https://apidocs.tally.xyz/){target=\_blank}.

> ## Overview

Reviewer 2: The title is also very general since your doing a comparison here.
```suggestion
##  Staking and Voting Comparison
```

> |                     | EVM (ETH, OP, ARB, BASE)                          | Solana                                         |

Reviewer 2: I want to say a title should be used for the first column,

Reviewer 2: A title should be added here, since it will help better organize the information.

Reviewer 2: https://developers.google.com/tech-writing/one/lists-and-tables?utm_source=chatgpt.com#create_useful_tables
It's one of the rules mentioned in the G- style guide that we try to follow.

|---------------------|---------------------------------------------------|------------------------------------------------|
| **Staking**         | Delegate your W tokens to another address.<br>Tokens stay in your wallet. | Move W tokens into a stake account and assign a delegate. |
| **Unstaking**       | Delegation removed; no token movement.                   | Withdraw tokens from stake account.     |

> | **Partial Staking** | Not supported (delegation applies to your full balance). | Supported; stake any amount of W.       |

Reviewer 2: I was thinking about this, what is Partial staking, and why is it relevant to this page.  Later on information about staking and unstaking is given but I don't know anything about partial. Can more be added for this?

Author: Partial staking is the same as staking but you just stake an amount that you decide instead of the full balance, machanics are the same as staking its just the amount that changes - no need to add a dedicated section on partial staking

Author: on evm the staked tokens remain in your wallet, so you are staking the full balance, you can't divide it, while for solana you send the amount of tokens u want to stake to a delegator so you choose the amount and don't have to send your full balance

> | **Voting Power**    | The delegate gets full voting rights for your W balance. | The delegate gets voting rights only for the amount staked. |

Reviewer 2: The previous three involve staking but why is voting power included here? Should this be fleshed out more in this document in its relation to W staking since its included to the table?

Author: voting power is tied to the amount you stake/delegate, the more u delegate the more voting power you give to ur delegator - these are pretty basic concepts of staking in blockchain, so I don't think voting needs a parahraph

## EVM Integration (ETH, OP, ARB, BASE)

> On EVM chains, staking is done by delegating your W balance to a delegate address using a single contract call. Tokens remain in your wallet, and delegation applies to your entire W balance.

Reviewer 2: ```suggestion
On EVM chains, staking means delegating your W balance to a delegate address in one contract call. The tokens stay in your wallet, but delegation covers your entire balance.
```

### W Token Contract Call

```js 

> const tx = await writeContractAsync({

Reviewer 2: Where does writeContractAsync come from as well?

Reviewer 2: For the other import, it comes from the things mentioned in the prereqs, but from at least my searching this one doesn't is this a standard in the bare bones EVM ecosystem ( I could be wrong) or is there another place this one comes from? The Solana section has the prereq section but this one for the EVM doesn't have that sort of supportive explanation.

  address: W_TOKEN_ADDRESS,
  abi: ERC1967,
  functionName: 'delegate',
  args: [isStaking ? delegateAddress : zeroAddress],
  chainId: chainId,
})
```

- To stake, pass the `delegateAddress`.
- To unstake, pass the `zeroAddress` (`0x000...000`).

Use the `delegates(address)` view function to check the current delegate for a given address.

!!!note "Important Notes"
    - There can only be one delegate per wallet. 
    - You can delegate to yourself or someone else.
    - Wormhole maintains a list of delegates using the Tally API (covers EVM and Solana).

## Solana Integration

> On Solana, staking means moving W into a stake custody account that’s tied to the staker and assigning a delegate for the voting power of the amount you staked. You can stake any amount and you can delegate to yourself or to a third party.

Reviewer 2: ```suggestion
On Solana, staking means moving W into a stake custody account that’s tied to the staker and assigning a delegate for the voting power of the amount you staked. Any amount can be staked, and the delegate can be yourself or another party.
```

### Prerequisites

- `@solana/web3.js`, `@solana/spl-token`, and `@coral-xyz/anchor` installed.

> - A Connection to the target cluster (e.g., mainnet).

Reviewer 2: ```suggestion
- A connection to the target cluster (e.g., mainnet).
```

- The W token mint on Solana.
- The [Staking program ID](#program-ids-and-abis){target=\_blank} (Anchor program) and its IDL.

> ### Inputs to  provide

Reviewer 2: ```suggestion
### Inputs to  Provide
```

> | Input.       | Type         | Description                                     | Example                                                 |

Reviewer 2: ```suggestion
| Input        | Type         | Description                                     | Example                                                 |
```

| `connection` | `Connection` | An RPC connection to the target Solana cluster. | `new Connection('https://api.mainnet-beta.solana.com')` |
| `wallet` | `AnchorProvider.wallet` | The wallet that signs transactions and pays fees. | `provider.wallet` |
| `programId` | `PublicKey` | The deployed staking program ID. | `new PublicKey('MGoV9M6YUsdhJzjzH9JMCW2tRe1LLxF1CjwqKC7DR1B')` |
| `idl` | `Idl` | The IDL for the staking program. | Loaded JSON file |
| `wMint` | `PublicKey` | The W token mint address on Solana. | `new PublicKey('85VBFQZC9TZkfaptBWjvUw7YbZjy52A6mjtPGjstQAmQ')` |
| `userPublicKey` | `PublicKey` | The staker’s wallet public key. | `wallet.publicKey` |
| `delegateePublicKey` | `PublicKey` | The address of the delegate who will receive voting power. | `new PublicKey('<DELEGATE_ADDRESS>')` |
| `currentDelegatePublicKey` | `PublicKey` | Your current delegate’s address when switching delegates (or `null` if staking for the first time). | `null` |
| `amount` | `BN` or `bigint` | The number of W tokens to stake in base units (respecting the token’s decimals). | `new BN(10 * 10 ** 9)` if W has 9 decimals |

### Staking Flow (Solana)

> Initialize the staking program and config. See [Program IDs and ABIs section](/docs/protocol/staking/w-staking/#program-ids-and-abis) for more details.

Reviewer 2: ```suggestion
Initialize the staking program and config. See [Program IDs and ABIs section](#program-ids-and-abis){target=\_blank} for more details.
```

---

# W Staking Integration

W staking allows users to stake their W tokens to participate in governance and earn staking rewards in return, while retaining control over their tokens. This guide walks you through integrating native W staking for both EVM chains (Ethereum, Optimism, Arbitrum, Base) and Solana, including high-level differences, contract calls, and recommended delegate discovery via the Tally API.

## Overview

|                     | EVM (ETH, OP, ARB, BASE)                          | Solana                                         |
|---------------------|---------------------------------------------------|------------------------------------------------|
| **Staking**         | Delegate your W tokens to another address.<br>Tokens stay in your wallet. | Move W tokens into a stake account and assign a delegate. |
| **Unstaking**       | Delegation removed; no token movement.                   | Withdraw tokens from stake account.     |
| **Partial Staking** | Not supported (delegation applies to your full balance). | Supported; stake any amount of W.       |

> | **Voting Power**    | The delegate gets full voting rights for your W balance. | The delegate gets voting rights only for the staked amount. |

Reviewer 2: ```suggestion
| **Voting Power**    | The delegate gets full voting rights for your W balance. | The delegate gets voting rights only for the staked amount. |
```

Author: what changed here?

Reviewer 2: ```suggestion
| **Voting Power**    | The delegate gets full voting rights for your W balance. | The delegate gets voting rights only for the amount staked. |
```
Sorry, just the quick swap between amount and staked. After this its good.

On EVM chains, staking is done by delegating your W balance to a delegate address using a single contract call. Tokens remain in your wallet, and delegation applies to your entire W balance.

### W Token Contract Call

```js 
const tx = await writeContractAsync({
  address: W_TOKEN_ADDRESS,
  abi: ERC1967,
  functionName: 'delegate',
  args: [isStaking ? delegateAddress : zeroAddress],
  chainId: chainId,
})
```

- To stake, pass the `delegateAddress`.
- To unstake, pass the `zeroAddress` (`0x000...000`).

Use the `delegates(address)` view function to check the current delegate for a given address.

!!!note "Important Notes"
    - There can only be one delegate per wallet. 
    - You can delegate to yourself or someone else.
    - Wormhole maintains a list of delegates using the Tally API (covers EVM and Solana).

## Solana Integration

On Solana, staking means moving W tokens into a stake account and assigning a delegate. You can stake any amount and delegate to any valid account (including yourself).

### Staking Flow (Solana)

Initialize the staking program and config. See [Program IDs and ABIs section](/docs/protocol/staking/w-staking/#program-ids-and-abis) for more details.

> ```js   

Reviewer: This could go inside the snippets folder

      stakeAccountMetadata: delegateeStakeAccountMetadata,
      custodyAuthority: delegateeCustodyAuthority,
      mint: WTokenSolanaPublicKey,
      stakeAccountCustody: delegateeStakeAccountCustody,
      tokenProgram: TOKEN_PROGRAM_ID,
      systemProgram: anchor.web3.SystemProgram.programId,
    })
    .instruction()
  createStakeAccTx.add(createDelegateeStakeIx)
}

// 5. Send delegate instructions
const delegateIx = await program.methods
  .delegate(delegateePublicKey, currentDelegatePublicKey)
  .accountsStrict({
    payer: walletProvider.publicKey,
    currentDelegateStakeAccountCheckpoints,
    currentDelegateStakeAccountMetadata,
    delegateeStakeAccountCheckpoints,
    delegateeStakeAccountMetadata,
    stakeAccountMetadata,
    custodyAuthority,
    stakeAccountCustody,
    vestingConfig: null,
    vestingBalance: null,
    config,
    mint: WTokenSolanaPublicKey,
    systemProgram: anchor.web3.SystemProgram.programId,
    eventAuthority,
    program: program.programId,
  })
  .instruction()
createStakeAccTx.add(delegateIx)
```

Staking steps:

1. Ensure the user has a W token ATA (Associated Token Account).

> 2. Create a stake account if it doesn't exist.

Reviewer 2: ```suggestion
2. Create a stake account if one doesn't already exist.
```

---
title: W Staking
description: Integrate W staking into your app on EVM and Solana. Learn flows, contract calls, and delegate discovery using the Tally API.
---

# W Staking Integration

> This guide walks through integrating native W staking for both EVM chains (Ethereum, Optimism, Arbitrum, Base) and Solana, including high-level differences, contract calls, and recommended delegate discovery via the Tally API.

Reviewer 2: ```suggestion
This guide walks you through integrating native W staking for both EVM chains (Ethereum, Optimism, Arbitrum, Base) and Solana, including high-level differences, contract calls, and recommended delegate discovery via the Tally API.
```

## Overview

|                       | EVM (ETH, OP, ARB, BASE)                          | Solana                                         |
|-----------------------|---------------------------------------------------|------------------------------------------------|
| **How staking works** | Delegate your W tokens to another address.<br>Tokens stay in your wallet. | Move W tokens into a stake account and assign a delegate. |

> | **Unstaking**         | Remove the delegate. Tokens don’t move.           | Withdraw tokens from stake account.            |

Reviewer 2: ```suggestion
| **Unstaking**         |  Delegation removed; no token movement.           | Withdraw tokens from stake account.            |
```

> | **Partial staking**   | Not supported (delegation applies to full balance). | Supported. Stake any amount of W.           |

Reviewer 2: ```suggestion
| **Partial staking**   | Not supported (delegation applies to your full balance). | Supported; stake any amount of W.           
```

> | **Voting Power**      | Delegate gets full voting rights for your W balance. | Delegate gets voting rights only for staked amount. |

Reviewer 2: ```suggestion
| **Voting Power**      | The delegate gets full voting rights for your W balance. | The delegate gets voting rights only for the staked amount. |
```

## EVM Integration (ETH, OP, ARB, BASE)

On EVM chains, staking is done by delegating your W balance to a delegate address using a single contract call. Tokens remain in your wallet, and delegation applies to your entire W balance.

### W Token Contract Call

```js 
const tx = await writeContractAsync({
  address: W_TOKEN_ADDRESS,
  abi: ERC1967,
  functionName: 'delegate',
  args: [isStaking ? delegateAddress : zeroAddress],
  chainId: chainId,
})
```

> - To stake, pass the delegate address.

Reviewer 2: Should delegate address be in back ticks and in the parameter from like delegateAddress since, I think, its referring to the args of the above code? Same with Zero Address?

- To unstake, pass the zero address (`0x000...000`).

Use the `delegates(address)` view function to check the current delegate for a given address.

!!!note "Important Notes"

>     - Only one delegate per wallet.

Reviewer 2: ```suggestion
    - There can only be one delegate per wallet. 
```
This flows better I think .

    - You can delegate to yourself or someone else.
    - Wormhole maintains a list of delegates using the Tally API (covers EVM and Solana).

## Solana Integration

On Solana, staking means moving W tokens into a stake account and assigning a delegate. You can stake any amount and delegate to any valid account (including yourself).

### Staking Flow (Solana)

```js 

>  // Initialize staking program and config

Reviewer 2: ```suggestion
 // Initialize the staking program and config
```

    [Buffer.from('authority'), userPublicKey.toBuffer()],
    program.programId,
  )

  const [stakeAccountCustody] = PublicKey.findProgramAddressSync(
    [Buffer.from('custody'), userPublicKey.toBuffer()],
    program.programId,
  )

  const [eventAuthority] = PublicKey.findProgramAddressSync(
    [Buffer.from('__event_authority')],
    program.programId,
  )

  const [currentDelegateStakeAccountMetadata] =
    PublicKey.findProgramAddressSync(
      [Buffer.from('stake_metadata'), currentDelegatePublicKey.toBuffer()],
      program.programId,
    )

  const [currentDelegateStakeAccountCheckpoints] =
    PublicKey.findProgramAddressSync(
      [
        Buffer.from('owner'),
        currentDelegatePublicKey.toBuffer(),
        Buffer.from([0, 0]),
      ],
      program.programId,
    )

  const [delegateeStakeAccountMetadata] = PublicKey.findProgramAddressSync(
    [Buffer.from('stake_metadata'), delegateePublicKey.toBuffer()],
    program.programId,
  )

  const [delegateeStakeAccountCheckpoints] = PublicKey.findProgramAddressSync(
    [Buffer.from('owner'), delegateePublicKey.toBuffer(), Buffer.from([0, 0])],
    program.programId,
  )
  

> // 1. Ensure ATA exists for user

Reviewer 2: ```suggestion
// 1. Ensure ATA exists for the user
```

const userATA = getAssociatedTokenAddressSync(WTokenSolanaPublicKey, userPublicKey, false)
const ataInfo = await connection.getAccountInfo(userATA)
if (!ataInfo) {
  createStakeAccTx.add(
    createAssociatedTokenAccountInstruction(
      walletProvider.publicKey,
      userATA,
      userPublicKey,
      WTokenSolanaPublicKey
    )
  )
}

> // 2. Create user's stake account if needed

Reviewer 2: ```suggestion
// 2. Create the user's stake account if needed
```

  
const userStakeAccountInfo = await connection.getAccountInfo(stakeAccountMetadata)
if (!userStakeAccountInfo) {
  const createUserStakeIx = await program.methods
    .createStakeAccount()
    .accountsStrict({
      payer: walletProvider.publicKey,
      config,
      stakeAccountCheckpoints,
      stakeAccountMetadata,
      custodyAuthority,
      mint: WTokenSolanaPublicKey,
      stakeAccountCustody,
      tokenProgram: TOKEN_PROGRAM_ID,
      systemProgram: anchor.web3.SystemProgram.programId,
    })
    .instruction()
  createStakeAccTx.add(createUserStakeIx)
}

// 3. Transfer tokens to custody

const transferIx = createTransferInstruction(userATA, stakeAccountCustody, userPublicKey, amount)
createStakeAccTx.add(transferIx)

> // 4. Create delegate account if not self-delegation

Reviewer 2: ```suggestion
// 4. Create a delegate account if not self-delegation
```

  
const delegateeAccountInfo = await connection.getAccountInfo(delegateeStakeAccountMetadata)
const isSelfDelegation = delegateePublicKey.equals(userPublicKey)
if (!delegateeAccountInfo && !isSelfDelegation) {
  const [delegateeCustodyAuthority] = PublicKey.findProgramAddressSync([
    Buffer.from('authority'),
    delegateePublicKey.toBuffer()
  ], program.programId)

  const [delegateeStakeAccountCustody] = PublicKey.findProgramAddressSync([
    Buffer.from('custody'),
    delegateePublicKey.toBuffer()
  ], program.programId)

  const createDelegateeStakeIx = await program.methods
    .createStakeAccount()
    .accountsStrict({
      payer: walletProvider.publicKey,
      config,
      stakeAccountCheckpoints: delegateeStakeAccountCheckpoints,
      stakeAccountMetadata: delegateeStakeAccountMetadata,
      custodyAuthority: delegateeCustodyAuthority,
      mint: WTokenSolanaPublicKey,
      stakeAccountCustody: delegateeStakeAccountCustody,
      tokenProgram: TOKEN_PROGRAM_ID,
      systemProgram: anchor.web3.SystemProgram.programId,
    })
    .instruction()
  createStakeAccTx.add(createDelegateeStakeIx)
}

> // 5. Send delegate instruction

Reviewer 2: ```suggestion
// 5. Send delegate instructions
```

const delegateIx = await program.methods
  .delegate(delegateePublicKey, currentDelegatePublicKey)
  .accountsStrict({
    payer: walletProvider.publicKey,
    currentDelegateStakeAccountCheckpoints,
    currentDelegateStakeAccountMetadata,
    delegateeStakeAccountCheckpoints,
    delegateeStakeAccountMetadata,
    stakeAccountMetadata,
    custodyAuthority,
    stakeAccountCustody,
    vestingConfig: null,
    vestingBalance: null,
    config,
    mint: WTokenSolanaPublicKey,
    systemProgram: anchor.web3.SystemProgram.programId,
    eventAuthority,
    program: program.programId,
  })
  .instruction()
createStakeAccTx.add(delegateIx)
```

Staking Steps:

> 1. Ensure user has a W token ATA.

Reviewer 2: ```suggestion
1. Ensure the user has a W token ATA.
```

> 2. Create stake account if it doesn't exist.

Reviewer 2: ```suggestion
2. Create a stake account if it doesn't exist.
```

3. Transfer tokens to the stake custody account.

> 4. Create delegate stake account if needed.

Reviewer 2: ```suggestion
4. Create a delegate stake account if needed.
```

5. Send the `delegate()` instruction.

### Unstaking Flow (Solana)

```js 
const withdrawIx = await program.methods
.withdrawTokens(
    new BN(amount.toString()),
    delegateePublicKey, // current delegate (who owns your stake)

>     userPublicKey, // withdraw back to yourself

Reviewer 2: // withdraw back to yourself sounds a bit weird to me, is it possible to change it? Maybe to the below or something similar depending on what you are trying to convey. 
```suggestion
    userPublicKey, // recipient of the withdrawn tokens
```

)
.accountsStrict({
    payer: walletProvider.publicKey,
    currentDelegateStakeAccountCheckpoints,
    currentDelegateStakeAccountMetadata,
    destination: userATA,
    stakeAccountMetadata,
    stakeAccountCustody,
    custodyAuthority,
    config,
    tokenProgram: TOKEN_PROGRAM_ID,
    systemProgram: anchor.web3.SystemProgram.programId,
    eventAuthority,
    program: program.programId,
})
.instruction()

withdrawTx.add(withdrawIx)
```

> 1. Confirm user’s stake metadata and custody accounts.

Reviewer 2: ```suggestion
1. Confirm the user’s stake metadata and custody accounts.
```

> 2. Call `withdrawTokens()` to move tokens back to user’s ATA.

Reviewer 2: ```suggestion
2. Call `withdrawTokens()` to move tokens back to the user’s ATA.
```

!!!note "Important Notes"
    - Solana integration uses `@solana/web3.js`, `@solana/spl-token`, and `@coral-xyz/anchor`.
    - Delegating to an active voter is generally recommended (see [Tally API](https://apidocs.tally.xyz/){target=\_blank}).
    - Both self- and third-party delegation are supported.

## Using the Tally API

Wormhole leverages the [Tally API](https://apidocs.tally.xyz/){target=\_blank} to fetch live delegate and voting data across both EVM and Solana.

- To fetch Solana delegates, use the Program ID as the `governorId`.
- You will need your Tally `organizationId`.

## Program IDs and ABIs

**EVM**

- W Token Contract: `0xb0ffa8000886e57f86dd5264b9582b2ad87b2b91`
- ERC1967 delegate ABI ([Etherscan link](https://etherscan.io/address/0xb0ffa8000886e57f86dd5264b9582b2ad87b2b91#readProxyContract){target=\_blank})

**Solana**

- W Token Address: `85VBFQZC9TZkfaptBWjvUw7YbZjy52A6mjtPGjstQAmQ`
- Current MultiGov Program ID: `MGoV9M6YUsdhJzjzH9JMCW2tRe1LLxF1CjwqKC7DR1B`
- IDL: [Solscan IDL](https://solscan.io/account/MGoV9M6YUsdhJzjzH9JMCW2tRe1LLxF1CjwqKC7DR1B#anchorProgramIdl){target=\_blank}

> For Anchor Program Integration (like the sample code above): You need the IDL file

Reviewer 2: ```suggestion
To integrate with an Anchor program, you need the IDL file (as shown in the sample code above) : 
```
When rendered I think this will look better

> - Get it from: [Solscan IDL](https://solscan.io/account/MGoV9M6YUsdhJzjzH9JMCW2tRe1LLxF1CjwqKC7DR1B#anchorProgramIdl)

Reviewer 2: ```suggestion
- You can retrieve it from: [Solscan IDL](https://solscan.io/account/MGoV9M6YUsdhJzjzH9JMCW2tRe1LLxF1CjwqKC7DR1B#anchorProgramIdl){target=\_blank}
```

> - Used in:

Reviewer 2: ```suggestion
- As seen in:
```

```js
const program = new anchor.Program<Staking>(simplerStakingIDL, {
  connection,
})
```

> For PDA derivations, Tally API integration, or custom integrations you will need the Program ID. Example PDA derivation use case:

Reviewer 2: ```suggestion
For PDA derivations, Tally API integration, or custom integrations, you will need the Program ID. Example PDA derivation use case:
```

```js
// If you already have the IDL/program above, use program.programId
const programId = new PublicKey(
    'MGoV9M6YUsdhJzjzH9JMCW2tRe1LLxF1CjwqKC7DR1B',
)
const [pda] = PublicKey.findProgramAddressSync(
    [Buffer.from('stake_metadata'), wallet.toBuffer()],
    programId,
)
```

> Example Tally API call using ProgramID:

Reviewer 2: ```suggestion
Implementation of a Tally API call using ProgramID:
```
